larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
-
-    if (bitpacked_input_shape.Dims(3) % 4 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 4) {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, true>, 8, 4, 4};
-      } else {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, false>, 8, 4, 4};
-      }
-    } else if (bitpacked_input_shape.Dims(3) % 2 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 2) {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, true>, 8, 4, 2};
-      } else {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, false>, 8, 4, 2};
-      }
-    } else {
-      if (bitpacked_input_shape.Dims(3) > 1) {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, true>, 8, 4, 1};
-      } else {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, false>, 8, 4, 1};
+  const std::int32_t input_depth;
+  const std::int32_t output_channels;
+  const std::int32_t filter_size;
+  const std::int32_t groups;
+  const std::int32_t num_output_pixels;
+
+  std::vector<TBitpacked> packed_weights;
+  std::vector<const TBitpacked*> indirection_buffer;
+  std::vector<TBitpacked> zero_buffer;
+
+  Kernel(const std::int32_t block_size_output_channels,
+         const std::int32_t block_size_pixels,
+         const std::int32_t block_size_depth,
+         const bconv2d::BConv2DParams* bconv2d_params,
+         const RuntimeShape& bitpacked_input_shape,
+         const RuntimeShape& output_shape)
+      : block_size_output_channels(block_size_output_channels),
+        block_size_pixels(block_size_pixels),
+        block_size_depth(block_size_depth),
+        input_depth(bitpacked_input_shape.Dims(3)),
+        output_channels(bconv2d_params->channels_out),
+        filter_size(bconv2d_params->filter_height *
+                    bconv2d_params->filter_width),
+        groups(bconv2d_params->groups),
+        num_output_pixels(bitpacked_input_shape.Dims(0) * output_shape.Dims(1) *
+                          output_shape.Dims(2)) {}
+
+  /**
+   * Pack the weights in the correct order. This procedure is (heavily)
+   * adapted from the following XNNPack function:
+   * https://github.com/google/XNNPACK/blob/80a8ac59849bfdae8d2e1409f5642baa502c0b9e/src/packing.c#L429-L484
+   */
+  void PackWeights(const TBitpacked* weights_ptr) {
+    const std::int32_t input_depth_per_group = input_depth / groups;
+    const std::int32_t output_channels_per_group = output_channels / groups;
+    const std::int32_t rounded_up_output_channels_per_group =
+        block_size_output_channels *
+        ((output_channels_per_group + block_size_output_channels - 1) /
+         block_size_output_channels);
+    packed_weights.resize(groups * rounded_up_output_channels_per_group *
+                              filter_size * input_depth_per_group +
+                          /* padding */ block_size_output_channels *
+                              block_size_depth);
+    std::int32_t packed_weights_index = 0;
```

Comment:
Here and elsewhere in this struct: is it really needed to specify `std::int32_t`, i.e. does it need to be 32-bits specifically? Typically we would use `int` if we just want a system-default integer. Also, if we do care so much, doesn't it make sense to use `uint32_t` for some of the counters (this would normally be a `size_t`)?

Another alternative is to use `auto` in more places to avoid (unnecessary) conversion or re-specifications of types.

Reply:
Yeah that's a fair point. I instinctively use `int32_t` by default for almost all C code I write because I think it's dumb that the default integer type could randomly be 16 bits on some targets. E.g. with Rust you almost always use explicit integer types `i8`, `i16`, `i32`, `i64`, et cetera.

Reply:
Agreed. So please do use `int32_t` if it is needed, but for most places where it doesn't matter that much (e.g. a counter from 0 till the filter width) I guess a regular `int` is better suited?

Reply:
For what its worth, I think the common thing to use in C++ for 'lenghts of arrays' and array indices is to use `size_t` which is unsigned and pointer-sized, i.e. 64-bit on 64-bit systems, to ensure you can exactly index everything that you could point potentially point to. An `int` is often 32-bit on 64-bit systems so then you can't index anything outside of a 4 GB array.
In this particular case I don't think it matters.